pytket/extensions/cuquantum/mps/mps.py

```diff
@@ -0,0 +1,588 @@
+        """Return the unique identifiers of the virtual bonds
+        of the tensor ``tensors[position]``.
+
+        Args:
+            position: A position in the MPS.
+
+        Returns:
+            A list with the ID of the virtual bonds of the tensor
+            in order from left to right.
+            If ``position`` is the first or last in the MPS, then the list
+            will only contain the corresponding virtual bond.
+        """
+        if position < 0 or position >= len(self):
+            raise Exception(f"Position {position} is out of bounds.")
+        elif position == 0:
+            v_bonds = [position + 1]
+        elif position == len(self) - 1:
+            v_bonds = [position]
+        else:
+            v_bonds = [position, position + 1]
+
+        assert all(vb in self.tensors[position].bonds for vb in v_bonds)
+        return v_bonds
+
+    def get_virtual_dimensions(self, position: int) -> list[int]:
+        """Return the dimension of the virtual bonds of the tensor
+        ``tensors[position]``.
+
+        Args:
+            position: A position in the MPS.
+
+        Returns:
+            A list with the dimensions of the virtual bonds of the
+            tensor in order from left to right.
+            If ``position`` is the first or last in the MPS, then the list
+            will only contain the corresponding virtual bond.
+        """
+        return [
+            self.tensors[position].get_dimension_of(bond)
+            for bond in self.get_virtual_bonds(position)
+        ]
+
+    def get_physical_bond(self, position: int) -> Bond:
+        """Return the unique identifier of the physical bond of the tensor
+        ``tensors[position]``.
+
+        Args
+            position: A position in the MPS.
+
+        Returns:
+            The identifier of the physical bond.
+        """
+        if position < 0 or position >= len(self):
+            raise Exception(f"Position {position} is out of bounds.")
+
+        # By construction, the largest identifier is the physical one
+        return max(self.tensors[position].bonds)
+
+    def get_physical_dimension(self, position: int) -> int:
+        """Return the dimension of the physical bond of the tensor
+        ``tensors[position]``.
+
+        Args:
+            position: A position in the MPS.
+
+        Returns:
+            The dimension of the physical bond.
+        """
+        return self.tensors[position].get_dimension_of(self.get_physical_bond(position))
+
+    def copy(self) -> MPS:
+        """
+        Returns:
+            A deep copy of the MPS
+        """
+        self._flush()
+
+        # Create object without initialising to |0> state
+        new_mps = MPS(n_tensors=-1, chi=self.chi)
+        # Copy all data
+        new_mps.fidelity = self.fidelity
+        new_mps.tensors = [t.copy() for t in self.tensors]
+        new_mps._complex_t = self._complex_t
+        new_mps._real_t = self._real_t
+
+        return new_mps

```

Comment:
I've finally made my mind about this. Since we intend to eventually have a general class (I'll call it `GeneralTN` here, for now) for tensor networks that captures both the current `TensorNetwork` and `MPS`, I think having a dagger method (and possibly, others such as `vdot`) is `GeneralTN`'s responsibility.

I'd say that `MPS` should provide methods to export and import its data as a `GeneralTN` object and `MPS` should be used exclusively to work with MPS algorithms.

I don't think think `MPS` should inherit from `GeneralTN`. It's fine to separate concerns here so that the data structure in MPS is fine tuned for MPS algorithms without overloading the definition of `GeneralTN`. I see `GeneralTN` as a transparent used for I/O with the user and for basic utils such as taking daggers and inner products, while `MPS` is more like a black box that you'll often call via `simulate` or such sort of high level API.

Reply:
This makes sense due to the way MPS is used in the simulate algorithm and is not exposed to the user.  Canonicalisation is unique to MPS. However, this could be a bit cumbersome if the user wanted to do vdots etc with an MPS. But if this is not the desired behaviour, then I agree with your decision. If we were building an arbitrary tensor network library I would disagree. However, having to define the cutensornet descriptors in multiple places requires extra code that could be stored in a base class.

After looking into the cutensornet Network class a bit more. If we rely on this for the circuits and exact contraction, which I think we should. Much of the lower-level functionality is wrapped up inside the Network class. If we use the Network class we would have to pull out the required attributes for the GeneralTN. So the workflow here is potentially a bit convoluted, however not impossible. This could be similar to what you suggest for MPS. The cutensor.Network class does not have any notion of vdots or joining networks so I think having this in a GeneralTN class would be useful.

Reply:
I am not convinced we should generalise different TN classes tbh. If we do though, we'll have to use inheritance.

Reply:
The other thing is that we could have a class or two to encapsulate cuTN bolierplate low-level code (like setting attributes to SVD). On the other hand, I do believe we should have a general Tensor class - perhaps piggy-backed from the one @PabloAndresCQ already has.